cpp/src/quantiles/quantiles.cu

```diff
@@ -16,196 +16,253 @@
 
 //Quantile (percentile) functionality
 
-#include "quantiles.h"
+#include "quantiles.hpp"
 #include "utilities/cudf_utils.h"
 #include "utilities/error_utils.hpp"
 #include "utilities/type_dispatcher.hpp"
+#include "utilities/wrapper_types.hpp"
 #include "rmm/thrust_rmm_allocator.h"
 #include "cudf.h"
 
 #include <thrust/device_vector.h>
 #include <thrust/copy.h>
 
-namespace{ //unknown
-  template<typename VType,
-           typename RetT = double>
-    void f_quantile_tester(rmm::device_vector<VType>& d_in)
+
+namespace{ // anonymouys namespace
+
+  struct QuantiledIndex {
+    gdf_size_type lower_bound;
+    gdf_size_type upper_bound;
+    gdf_size_type nearest;
+    double fraction;
+  };
+
+  QuantiledIndex find_quantile_index(gdf_size_type length, double quant)
   {
-    using FctrType = std::function<RetT(VType, VType, double)>;
+    // clamp quant value.
+    // Todo: use std::clamp if c++17 is supported.
+    quant = std::min(std::max(quant, 0.0), 1.0);
 
-    FctrType lin_interp{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(static_cast<double>(y0) + x*static_cast<double>(y1-y0));//(f(x) - y0) / (x - 0) = m = (y1 - y0)/(1 - 0)
-      }};
+    // since gdf_size_type is int32_t, there is no underflow/overflow
+    double val = quant*(length -1);
+    QuantiledIndex qi;
+    qi.lower_bound = std::floor(val);
+    qi.upper_bound = static_cast<size_t>(std::ceil(val));
+    qi.nearest = static_cast<size_t>(std::nearbyint(val));
+    qi.fraction = val - qi.lower_bound;
 
-    FctrType midpoint{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(static_cast<double>(y0 + y1)/2.0);
-      }};
+    return qi;
+  }
 
-    FctrType nearest{[](VType y0, VType y1, double x){
-        return static_cast<RetT>(x < 0.5 ? y0 : y1);
-      }};
+  template<typename T>
+  void singleMemcpy(T& res, T* input, cudaStream_t stream = NULL)

```

Comment:
Why do we need a function that just wraps a memcpy? This isn't really saving any typing, and it's local to this file so it's not a utility...

Reply:
I tried to delete this comment after I saw how much you use it, but Github wasn't cooperating. Sorry,.